src/bin/rotel/main.rs

```diff
@@ -145,29 +150,46 @@ async fn run_agent(
 ) -> Result<(), Box<dyn Error + Send + Sync>> {
     let mut agent_join_set = JoinSet::new();
 
+    let (mut flush_pipeline_tx, flush_pipeline_sub) = FlushBroadcast::new().into_parts();
+    let (mut flush_exporters_tx, flush_exporters_sub) = FlushBroadcast::new().into_parts();
+
     let cancel_token = CancellationToken::new();
     {
         let token = cancel_token.clone();
         let env = env.clone();
         let agent_fut = async move {
-            let agent = Agent::new(agent_args, port_map, SENDING_QUEUE_SIZE, env);
+            let agent = Agent::new(agent_args, port_map, SENDING_QUEUE_SIZE, env)
+                .with_pipeline_flush(flush_pipeline_sub)
+                .with_exporters_flush(flush_exporters_sub);
             agent.run(token).await
         };
 
         agent_join_set.spawn(agent_fut);
     };
 
-    select! {
-        _ = signal_wait() => {
-            info!("Shutdown signal received.");
-            cancel_token.cancel();
-        },
-        e = wait::wait_for_any_task(&mut agent_join_set) => {
-            match e {
-                Ok(()) => warn!("Unexpected early exit of agent."),
-                Err(e) => return Err(e),
-            }
-        },
+    let mut sig_usr1 = sig(SignalKind::user_defined1());
+    loop {
+        select! {
+            _ = signal_wait() => {
+                info!("Shutdown signal received.");
+                cancel_token.cancel();
+                break;
+            },
+            _ = sig_usr1.recv() => {
+                info!("Signal SIGUSR1 received, invoking a forced flush");
+                let flush_start = Instant::now();
+                force_flush(&mut flush_pipeline_tx, &mut flush_exporters_tx).await;
+                let duration = Instant::now().duration_since(flush_start);
+                info!(duration = ?duration, "Finished forced flush request");
```

Comment:
Should you break here?

Reply:
No, this is the one branch we don't want to break from since that would terminate the agent. After flush we want to go back to waiting.

Reply:
oh you don't intend to shutdown here, I get it.